In [1]:
import psycopg2
from datetime import timedelta
from sqlalchemy import create_engine
import psycopg2
import pandas as pd
import numpy as np

RANDOM_SEED = 42

C:\Users\Theo\AppData\Local\Programs\Python\Python37\lib\site-packages\numpy\_distributor_init.py:32: UserWarning: loaded more than 1 DLL from .libs:
C:\Users\Theo\AppData\Local\Programs\Python\Python37\lib\site-packages\numpy\.libs\libopenblas.IPBC74C7KURV7CB2PKT5Z5FNR3SIBV4J.gfortran-win_amd64.dll
C:\Users\Theo\AppData\Local\Programs\Python\Python37\lib\site-packages\numpy\.libs\libopenblas.XWYDX2IKJW2NMTWSFYNGFUWKQU3LYTCZ.gfortran-win_amd64.dll
  stacklevel=1)


## Read in data
Prerequisites:
1. Build postgres-functions (\i path_to_file/postgres-functions.sql)
2. Build flicu_icustay_detail (\i path_to_file/flicu_icustay_detail.sql)
3. Build pivoted_vital (\i path_to_file/pivoted_vital.sql)
4. Build flicu_pivoted_lab (\i path_to_file/flicu_pivoted_lab.sql) (alternatively, build pivoted_lab if lab values before ICU admission are needed)

In [ ]:
# Connect to db
conn = psycopg2.connect("dbname=mimic user=mimic3 password=mimic3") 
cur = conn.cursor() 

# Read in table with patients & admissions (inner join on subject_id) and icu_stays (inner joinon subject_id and hadm_id)
icustay_details = pd.read_sql_query("SELECT * FROM mimiciii.flicu_icustay_detail;", conn)

# Read in vital signs
pivoted_vital = pd.read_sql_query("SELECT * FROM mimiciii.pivoted_vital;", conn)

# Read in lab measurements
# Use flicu_pivoted_lab (as it only takes the lab tests during ICU stay)
query = "SELECT * FROM mimiciii.flicu_pivoted_lab;"
# Alternative:  Use the lab values recorded previous to the ICU stay (although during same hospital admission!), 
# then sample them  (8h intervalls) and then forward fill plus cap at either icu admission time of first vital sign recorded
#query = "SELECT * FROM mimiciii.pivoted_lab;"
pivoted_lab = pd.read_sql_query(query, conn)

# Close the cursor and connection to so the server can allocate bandwidth to other requests
cur.close()
conn.close()

In [ ]:
print(icustay_details.info())
print(icustay_details.head(3))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61051 entries, 0 to 61050
Data columns (total 23 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   subject_id                   61051 non-null  int64         
 1   hadm_id                      61051 non-null  int64         
 2   icustay_id                   61051 non-null  int64         
 3   gender                       61051 non-null  object        
 4   dod                          23889 non-null  datetime64[ns]
 5   admittime                    61051 non-null  datetime64[ns]
 6   dischtime                    61051 non-null  datetime64[ns]
 7   los_hospital                 61051 non-null  float64       
 8   admission_age                61051 non-null  float64       
 9   ethnicity                    61051 non-null  object        
 10  ethnicity_grouped            61051 non-null  object        
 11  hospital_expire_flag         61051 non-nu

TEST: It should be 46428 patients and 61051 admissions

In [ ]:
data_1 = icustay_details.copy()
print("Number of patients: ", data_1['subject_id'].nunique())
print("Number of icu stays/admissions: ", data_1['icustay_id'].nunique())

Number of patients:  46428
Number of icu stays/admissions:  61051


## Patient/ICU stay Filtering
1. Filter for the first ICU admission of each patient
2. Exclude patients from NICU/PICU
3. Filter for ICU stays that were at least 48h long
4. Exclude patients with data recorder for less thatn 48h

### 1. Filter for the first ICU admission of each patient
We included only the first admission of each patient in the ICU, which resulted in each patient having only one ICU admission.

In [ ]:
data_2 = data_1.copy()
#data_2 = data_2[data_2['first_icu_stay_current_hosp'] == True]
data_2 = data_2[data_2['first_icu_stay_patient'] == True]

Test: Each patient should have only one ICU stayd and its respecitve hospital admission

In [ ]:
print("Number of patients: ", data_2['subject_id'].nunique())
print("Number of admissions: ", data_2['hadm_id'].nunique())
print("Number of ICU stays: ", data_2['icustay_id'].nunique())

Number of patients:  46428
Number of admissions:  46428
Number of ICU stays:  46428


### 2. Exclude patients from NICU/PICU
Patients admitted to the neonatal intensive care unit (NICU) and pediatric
intensive care unit (PICU) were excluded.

In [ ]:
data_2 = data_2[data_2.first_careunit != "NICU"]
data_2 = data_2[data_2.first_careunit != "PICU"]

TEST: It should be 38566 patients/icu stays

In [ ]:
print("Number of patients: ", data_2['subject_id'].nunique())
print("Number of ICU stays: ",data_2['icustay_id'].nunique())

Number of patients:  38566
Number of ICU stays:  38566


In [ ]:
data_2[["intime", "outtime", "los_icu", "deathtime_icu", "label_death_icu"]].head(3)

intime             outtime   los_icu deathtime_icu  \
1 2101-10-20 19:10:11 2101-10-26 20:43:09  6.064560           NaT   
2 2191-03-16 00:29:31 2191-03-17 16:46:31  1.678472           NaT   
4 2175-05-30 21:30:54 2175-06-03 13:39:54  3.672917           NaT   

   label_death_icu  
1                0  
2                0  
4                0

### 3. Filter for ICU stays that were at least 48h long

In [ ]:
data_3 = data_2.copy()
data_3 = data_3[data_3.los_icu >= 2]

TEST: It should be 20396 patients/icu stays

In [ ]:
print("Number of patients: ", data_3['subject_id'].nunique())
print("Number of icu stays: ", data_3['icustay_id'].nunique())

Number of patients:  20396
Number of icu stays:  20396


Temporary results of filtering

In [ ]:
filtered_icustay_ids = pd.DataFrame(data_3['icustay_id'].unique(), columns=['icustay_id'])

### 4. Exclude patients with data recorded for less than 48h
Excluded patients for whom the duration between the first and last observations of vital signs and laboratory tests was less than 48h, i.e. first_recorded_value - intime <= 48h. The duration was calculated as the last timestamp minus the first timestamp in the chartevents/labevents table.

In [ ]:
# Drop measurements with no belonging icustay_id
pivoted_vital = pivoted_vital.dropna(subset=['icustay_id'])
pivoted_lab = pivoted_lab.dropna(subset=['icustay_id'])
# Drop all rows only containing NaN values
pivoted_vital = pivoted_vital.dropna(how='all')
pivoted_lab = pivoted_lab.dropna(how='all')
# Drop all duplicate rows
pivoted_vital = pivoted_vital.drop_duplicates()
pivoted_lab = pivoted_lab.drop_duplicates()

# Leave only relevant columns
vital_colums = ['icustay_id', 'charttime', 'heartrate', 'sysbp', 'diasbp', 'meanbp', 'resprate', 'tempc', 'spo2']
pivoted_vital = pivoted_vital[vital_colums]
lab_columns = ['icustay_id', 'charttime', 'albumin', 'bun', 'bilirubin', 'lactate', 'bicarbonate', 'bands', 'chloride', 'creatinine', 'glucose',
        'hemoglobin', 'hematocrit', 'platelet', 'potassium', 'ptt', 'sodium', 'wbc']
pivoted_lab = pivoted_lab[lab_columns]

# Cast icustay_id types to int
pivoted_vital['icustay_id'] = pivoted_vital['icustay_id'].astype(int)
pivoted_lab['icustay_id'] = pivoted_lab['icustay_id'].astype(int)

# Keep only values of patients in previously filtered icustay_ids in labs and vitals
# But take all filtered patients (also ones that e.g. don't have any lab values recorded)
pivoted_vital = pivoted_vital.merge(filtered_icustay_ids, on='icustay_id', how='right').drop_duplicates()
pivoted_lab = pivoted_lab.merge(filtered_icustay_ids, on='icustay_id', how='right').drop_duplicates()

Test: Number of ICU stays in vitals and labs should be the same as the filtered ones above - 20396

In [ ]:
print("Number of ICU stays in pivoted_vital_filtered: ", pivoted_vital['icustay_id'].nunique())
print("Number of ICU stays in pivoted_lab_filtered: ", pivoted_lab['icustay_id'].nunique())

Number of ICU stays in pivoted_vital_filtered:  20396
Number of ICU stays in pivoted_lab_filtered:  20396


In [ ]:
# Min of each lab and vitals
icustay_ids_charttime_min_lab = pivoted_lab[["icustay_id", "charttime"]][pivoted_lab.groupby("icustay_id")["charttime"].rank(ascending=1,method='dense') == 1]
icustay_ids_charttime_min_vital = pivoted_vital[["icustay_id", "charttime"]][pivoted_vital.groupby("icustay_id")["charttime"].rank(ascending=1,method='dense') == 1]
# Min of both combined
icustay_ids_charttime_min_vital_lab = pd.concat([icustay_ids_charttime_min_lab, icustay_ids_charttime_min_vital], ignore_index=True)
icustay_ids_charttime_min_vital_lab = icustay_ids_charttime_min_vital_lab[["icustay_id", "charttime"]][icustay_ids_charttime_min_vital_lab.groupby("icustay_id")["charttime"].rank(ascending=1,method='dense') == 1]

# Max of each lab and vitals
icustay_ids_charttime_max_lab = pivoted_lab[["icustay_id", "charttime"]][pivoted_lab.groupby("icustay_id")["charttime"].rank(ascending=0,method='dense') == 1]
icustay_ids_charttime_max_vital = pivoted_vital[["icustay_id", "charttime"]][pivoted_vital.groupby("icustay_id")["charttime"].rank(ascending=0,method='dense') == 1]
# Max of both combined
icustay_ids_charttime_max_vital_lab = pd.concat([icustay_ids_charttime_max_lab, icustay_ids_charttime_max_vital], ignore_index=True)
icustay_ids_charttime_max_vital_lab = icustay_ids_charttime_max_vital_lab[["icustay_id", "charttime"]][icustay_ids_charttime_max_vital_lab.groupby("icustay_id")["charttime"].rank(ascending=0,method='dense') == 1]

In [ ]:
# Find for which icustay_ids there exist at least 48h of data
icustay_ids_vital_lab_charttime_min_max = pd.concat([icustay_ids_charttime_max_vital_lab, icustay_ids_charttime_min_vital_lab], ignore_index=True)
time_delta_48_hours = timedelta(days=0, seconds=0, microseconds=0, milliseconds=0, minutes=0, hours=48, weeks=0)
is_time_diff_bigger_48_lab = icustay_ids_vital_lab_charttime_min_max.groupby(['icustay_id'])['charttime'].transform(lambda x: (x.max()-x.min())) >= time_delta_48_hours

icustay_ids_vital_lab_charttime_min_max_filtered = icustay_ids_vital_lab_charttime_min_max[is_time_diff_bigger_48_lab]
print("Unique icu stays in icustay_ids_vital_lab_charttime_min_max_filtered after filtering", icustay_ids_vital_lab_charttime_min_max_filtered['icustay_id'].nunique())

# Keep only icustay ids for which at least 48h of data exists
icustay_ids_time_filtered = pd.DataFrame(icustay_ids_vital_lab_charttime_min_max_filtered['icustay_id'].unique(), columns=['icustay_id'])
print("Unique icu stays in icustay_ids_time_filtered: ", icustay_ids_time_filtered['icustay_id'].nunique())

Unique icu stays in icustay_ids_vital_lab_charttime_min_max_filtered after filtering 19414
Unique icu stays in icustay_ids_time_filtered:  19414


#### Final set of filtered icustay ids (filtered_icustay_ids)

In [ ]:
filtered_icustay_ids = filtered_icustay_ids.merge(icustay_ids_time_filtered, on='icustay_id', how='inner').drop_duplicates()

Test: Final number of unique icustay ids should be 19414

In [ ]:
print("Unique icu stays (final): ", filtered_icustay_ids['icustay_id'].nunique())

Unique icu stays (final):  19414


#### Create subset of all datasets (pivoted_lab, pivoted_vital, demographics) based on all exclusion criteria

In [ ]:
demographics_filtered = data_3.merge(filtered_icustay_ids, on='icustay_id', how='right').drop_duplicates()
print("Number of ICU stays demographics: ", demographics_filtered['icustay_id'].nunique())

vital_filtered = pivoted_vital.merge(filtered_icustay_ids, on='icustay_id', how='right').drop_duplicates()
print("Number of ICU stays vitals: ", vital_filtered['icustay_id'].nunique())

lab_filtered = pivoted_lab.merge(filtered_icustay_ids, on='icustay_id', how='right').drop_duplicates()
print("Number of ICU stays labs: ", lab_filtered['icustay_id'].nunique())

Number of ICU stays demographics:  19414
Number of ICU stays vitals:  19414
Number of ICU stays labs:  19414


# DATA PREPARATION - ML format
Vital sign measurements were typically taken 0.5–1.5 times per hour for the MIMIC-III database, while laboratory measurements were typically taken 1–2 times per eight hours. Therefore, each vital sign variable was aggregated into a one-hour interval, whereas each laboratory variable was aggregated into an eight-hour interval. Repeated measurements in a single interval were aggregated by the median.

In [ ]:
# Observation: The lab values and vital signs don't have the same starting time
vital_filtered[["icustay_id", "charttime"]][vital_filtered["icustay_id"] == 226799].sort_values("charttime").head(3), lab_filtered[["icustay_id", "charttime"]][lab_filtered["icustay_id"] == 226799].sort_values("charttime").head(3)

(         icustay_id           charttime
 3092157      226799 2153-07-31 08:00:00
 3092158      226799 2153-07-31 08:09:00
 3092159      226799 2153-07-31 08:10:00,
         icustay_id           charttime
 277108      226799 2153-07-31 10:12:00
 277109      226799 2153-07-31 10:30:00
 277110      226799 2153-07-31 12:06:00)

### Align time of entries of Labs & Vitals
Make sure that the vital & lab measurements of each patient start and end at the same time (so that both input time frames are ending up in the same timeframe) - The code below adds the same time steps with NaN values.

In [ ]:
vital_filtered = vital_filtered.merge(lab_filtered[['icustay_id', 'charttime']], on=['icustay_id', 'charttime'], how='outer').drop_duplicates()
print("Number of ICU stays in lab_filtered: ", vital_filtered['icustay_id'].nunique())
lab_filtered = lab_filtered.merge(vital_filtered[['icustay_id', 'charttime']], on=['icustay_id', 'charttime'], how='outer').drop_duplicates()
print("Number of ICU stays in lab_filtered: ", lab_filtered['icustay_id'].nunique())

Number of ICU stays in lab_filtered:  19414
Number of ICU stays in lab_filtered:  19414


In [ ]:
# Test: Now both, lab measurements and vital signs, should start at the same time (additional rows with NaN values).
vital_filtered[vital_filtered["icustay_id"] == 226799].sort_values("charttime").head(3), lab_filtered[lab_filtered["icustay_id"] == 226799].sort_values("charttime").head(3)

(         icustay_id           charttime  heartrate  sysbp  diasbp  meanbp  \
 3092157      226799 2153-07-31 08:00:00        NaN    NaN     NaN     NaN   
 3092158      226799 2153-07-31 08:09:00        NaN   65.0    50.0    53.0   
 3092159      226799 2153-07-31 08:10:00      129.0    NaN     NaN     NaN   
 
          resprate      tempc  spo2  
 3092157       NaN  36.666667   NaN  
 3092158       NaN        NaN   NaN  
 3092159      28.0        NaN  98.0  ,
          icustay_id           charttime  albumin  bun  bilirubin  lactate  \
 3321913      226799 2153-07-31 08:00:00      NaN  NaN        NaN      NaN   
 3321914      226799 2153-07-31 08:09:00      NaN  NaN        NaN      NaN   
 3321915      226799 2153-07-31 08:10:00      NaN  NaN        NaN      NaN   
 
          bicarbonate  bands  chloride  creatinine  glucose  hemoglobin  \
 3321913          NaN    NaN       NaN         NaN      NaN         NaN   
 3321914          NaN    NaN       NaN         NaN      NaN         N

### Resample Vital Signs

In [ ]:
vital_resampled = vital_filtered.copy()

# Resample from the end of the time series
vital_resampled = vital_resampled.assign(charttime=vital_resampled.charttime.dt.round('H'))
vital_resampled = vital_resampled.set_index('charttime').groupby('icustay_id').resample('1H', origin="end").median().drop(['icustay_id'], axis = 1).reset_index()
# Resample from the beginning of the time series
#vital_resampled = vital_resampled.set_index('charttime').groupby('icustay_id').resample('1H', origin="start").median().drop(['icustay_id'], axis = 1).reset_index()

# Forward and backwards fill (use lambda function instead of directly applying it to groupby otherwise results from one group are carreid forward to another group...BAD)
# Fill NaNs (-1 or 0 or mean!?)
vital_col = vital_resampled.columns.drop(['icustay_id', 'charttime'])
vital_resampled = vital_resampled.set_index(['icustay_id', 'charttime']).groupby('icustay_id')[vital_col].transform(lambda x: x.ffill().bfill()).fillna(-1).reset_index()
#.fillna(value=vital_resampled[vital_columns].mean())#.fillna(0)#.fillna(-1)

print(vital_filtered[vital_filtered["icustay_id"]==229633].tail(9))
print(vital_resampled[vital_resampled["icustay_id"]==229633].tail(3))
print(vital_resampled.isnull().sum().sum())

# id 271896 has no entries in lab values, i.e. there will be NaN values filled with -1, mean, ... (whatever chosen)
vital_resampled[vital_resampled["icustay_id"]==271896]

         icustay_id           charttime  heartrate  sysbp  diasbp  meanbp  \
4368839      229633 2147-02-10 15:00:00       82.0  122.0    95.0   102.0   
4368840      229633 2147-02-10 16:00:00       81.0    NaN     NaN     NaN   
4368841      229633 2147-02-10 17:00:00       78.0  120.0    54.0    69.0   
4571964      229633 2147-02-08 15:07:00        NaN    NaN     NaN     NaN   
4571965      229633 2147-02-09 02:31:00        NaN    NaN     NaN     NaN   
4571966      229633 2147-02-09 07:35:00        NaN    NaN     NaN     NaN   
4571967      229633 2147-02-09 08:09:00        NaN    NaN     NaN     NaN   
4571968      229633 2147-02-09 15:46:00        NaN    NaN     NaN     NaN   
4571969      229633 2147-02-10 01:05:00        NaN    NaN     NaN     NaN   

         resprate  tempc  spo2  
4368839      21.0    NaN  97.0  
4368840      21.0    NaN   NaN  
4368841      20.0    NaN   NaN  
4571964       NaN    NaN   NaN  
4571965       NaN    NaN   NaN  
4571966       NaN    NaN   NaN 

icustay_id           charttime  heartrate  sysbp  diasbp  meanbp  \
2256597      271896 2147-04-28 11:00:00       -1.0   -1.0    -1.0    -1.0   
2256598      271896 2147-04-28 12:00:00       -1.0   -1.0    -1.0    -1.0   
2256599      271896 2147-04-28 13:00:00       -1.0   -1.0    -1.0    -1.0   
2256600      271896 2147-04-28 14:00:00       -1.0   -1.0    -1.0    -1.0   
2256601      271896 2147-04-28 15:00:00       -1.0   -1.0    -1.0    -1.0   
...             ...                 ...        ...    ...     ...     ...   
2256669      271896 2147-05-01 11:00:00       -1.0   -1.0    -1.0    -1.0   
2256670      271896 2147-05-01 12:00:00       -1.0   -1.0    -1.0    -1.0   
2256671      271896 2147-05-01 13:00:00       -1.0   -1.0    -1.0    -1.0   
2256672      271896 2147-05-01 14:00:00       -1.0   -1.0    -1.0    -1.0   
2256673      271896 2147-05-01 15:00:00       -1.0   -1.0    -1.0    -1.0   

         resprate  tempc  spo2  
2256597      -1.0   -1.0  -1.0  
2256598      -1.0   -1.0  -1.0  
2256599      -1.0   -1.0  -1.0  
2256600      -1.0   -1.0  -1.0  
2256601      -1.0   -1.0  -1.0  
...           ...    ...   ...  
2256669      -1.0   -1.0  -1.0  
2256670      -1.0   -1.0  -1.0  
2256671      -1.0   -1.0  -1.0  
2256672      -1.0   -1.0  -1.0  
2256673      -1.0   -1.0  -1.0  

[77 rows x 9 columns]

Test for classification - This must run error free for running the code later

In [ ]:
test = vital_resampled.copy()
test = test.groupby("icustay_id").head(48)
print(test.head(3))
print(test.groupby(["icustay_id"])["charttime"].nunique().unique())

   icustay_id           charttime  heartrate  sysbp  diasbp  meanbp  resprate  \
0      200003 2199-08-02 19:00:00      132.0  106.0    68.0    73.0      25.0   
1      200003 2199-08-02 20:00:00      119.0   91.0    49.0    58.0      35.0   
2      200003 2199-08-02 21:00:00      118.5   89.5    54.5    61.0      33.5   

       tempc  spo2  
0  39.555554  97.0  
1  39.555554  97.0  
2  38.999998  96.5  
[48]


### Resample Laboratory Measurements
Added bfill to make sure that those values are not NaN

In [ ]:
lab_resampled = lab_filtered.copy()
# Cut out minutes and hours, so that the resampling of the 8h takes the same time span as the 1h samples (for vitals)
lab_resampled = lab_resampled.assign(charttime=lab_resampled.charttime.dt.round('H'))
# Resample from the end of the time series
lab_resampled = lab_resampled.set_index('charttime').groupby('icustay_id').resample('8h', origin="end").median().drop(['icustay_id'], axis = 1).reset_index()
#lab_resampled = lab_resampled.set_index('charttime').groupby('icustay_id').resample('8h', origin="start").median().drop(['icustay_id'], axis = 1).reset_index()

# Forward and backwards fill (use transform instead of direct groupby otherwise results from one group are carreid forward to another group...BAD)
# Fill NaNs (-1 or 0 or mean!?)
lab_col = lab_resampled.columns.drop(['icustay_id', 'charttime'])
lab_resampled = lab_resampled.set_index(['icustay_id', 'charttime']).groupby('icustay_id')[lab_col].transform(lambda x: x.ffill().bfill()).fillna(-1).reset_index()

print(lab_filtered[lab_filtered["icustay_id"]==229633].tail(9))
print(lab_resampled[lab_resampled["icustay_id"]==229633].tail(3))
print(lab_resampled.isnull().sum().sum())

         icustay_id           charttime  albumin  bun  bilirubin  lactate  \
4571961      229633 2147-02-10 11:00:00      NaN  NaN        NaN      NaN   
4571962      229633 2147-02-10 11:27:00      NaN  NaN        NaN      NaN   
4571963      229633 2147-02-10 12:00:00      NaN  NaN        NaN      NaN   
4571964      229633 2147-02-10 13:00:00      NaN  NaN        NaN      NaN   
4571965      229633 2147-02-10 14:00:00      NaN  NaN        NaN      NaN   
4571966      229633 2147-02-10 14:01:00      NaN  NaN        NaN      NaN   
4571967      229633 2147-02-10 15:00:00      NaN  NaN        NaN      NaN   
4571968      229633 2147-02-10 16:00:00      NaN  NaN        NaN      NaN   
4571969      229633 2147-02-10 17:00:00      NaN  NaN        NaN      NaN   

         bicarbonate  bands  chloride  creatinine  glucose  hemoglobin  \
4571961          NaN    NaN       NaN         NaN      NaN         NaN   
4571962          NaN    NaN       NaN         NaN      NaN         NaN   
4571963

Test for classification  - This must run error free for running the code later

In [ ]:
test = lab_resampled.copy()
test = test.groupby("icustay_id").head(6)
print(test.head(3))
print(test.groupby(["icustay_id"])["charttime"].nunique().unique())

   icustay_id           charttime  albumin   bun  bilirubin  lactate  \
0      200003 2199-08-02 23:00:00      2.8  20.0        3.4      4.6   
1      200003 2199-08-03 07:00:00      2.8  20.0        3.4      4.6   
2      200003 2199-08-03 15:00:00      2.8  20.0        3.4      4.6   

   bicarbonate  bands  chloride  creatinine  glucose  hemoglobin  hematocrit  \
0         18.0   -1.0     105.0         1.0    159.0        10.8        35.0   
1         18.0   -1.0     105.0         1.0    159.0        10.8        35.0   
2         18.0   -1.0     105.0         1.0    159.0        10.8        35.0   

   platelet  potassium   ptt  sodium   wbc  
0     109.0        3.2  32.1   141.0  40.2  
1     109.0        3.2  32.1   141.0  40.2  
2     109.0        3.2  32.1   141.0  40.2  
[6]


## Labels
Patients who died during their ICU stay were identified by the deathtime variable in
the admission table of MIMIC-III.

Patients who died during their stay in the ICU were included in the positive group (output = 1), and patients who survived to discharge were included in the negative group (output = 0).

This is done as part of icustay_detail.sql and stored in demographics_filtered.

### Add label to vital and lab datasets

In [ ]:
final_vital = vital_resampled.merge(demographics_filtered[["icustay_id", "label_death_icu"]], on="icustay_id", how="right")
print("Number of ICU stays in final_vitals: ", final_vital['icustay_id'].nunique())

final_lab = lab_resampled.merge(demographics_filtered[["icustay_id", "label_death_icu"]], on="icustay_id", how="right")
print("Number of ICU stays in final_vitals: ", final_lab['icustay_id'].nunique())

Number of ICU stays in final_vitals:  19414
Number of ICU stays in final_vitals:  19414


In [ ]:
demographics_filtered["label_death_icu"].value_counts()

0    17522
1     1892
Name: label_death_icu, dtype: int64

### Some Statistical Information

Demogrpahics

In [ ]:
demographics_filtered.describe()

subject_id        hadm_id     icustay_id  los_hospital  \
count  19414.000000   19414.000000   19414.000000  19414.000000   
mean   36305.652879  149998.389461  250439.306583     13.456947   
std    28966.570968   28917.484208   28773.388017     12.313006   
min        3.000000  100001.000000  200003.000000      0.263889   
25%    12979.000000  124997.500000  225585.250000      6.281250   
50%    26149.500000  149812.000000  250885.500000      9.795139   
75%    59445.750000  175340.750000  275336.750000     16.301910   
max    99995.000000  199999.000000  299993.000000    294.660417   

       admission_age  hospital_expire_flag  hospstay_seq       los_icu  \
count   19414.000000          19414.000000       19414.0  19414.000000   
mean       74.975595              0.137118           1.0      6.817994   
std        53.212971              0.343980           0.0      7.500257   
min        15.186897              0.000000           1.0      2.000856   
25%        53.804782              0.000000           1.0      2.871270   
50%        67.014075              0.000000           1.0      4.070260   
75%        78.496017              0.000000           1.0      7.399604   
max       310.280861              1.000000           1.0    153.927975   

       icustay_seq  label_death_icu  
count      19414.0     19414.000000  
mean           1.0         0.097455  
std            0.0         0.296585  
min            1.0         0.000000  
25%            1.0         0.000000  
50%            1.0         0.000000  
75%            1.0         0.000000  
max            1.0         1.000000

In [ ]:
demographics_filtered[demographics_filtered["label_death_icu"]==1].describe()

subject_id        hadm_id     icustay_id  los_hospital  \
count   1892.000000    1892.000000    1892.000000   1892.000000   
mean   35017.714588  148820.245772  250655.385307     10.818916   
std    28363.112837   28915.368465   28754.870253     10.310810   
min        9.000000  100053.000000  200019.000000      0.263889   
25%    13119.750000  123569.000000  226370.500000      4.161458   
50%    24941.500000  148886.500000  251145.500000      7.469792   
75%    57379.250000  173636.500000  275780.250000     13.488194   
max    99955.000000  199919.000000  299895.000000     97.267361   

       admission_age  hospital_expire_flag  hospstay_seq      los_icu  \
count    1892.000000                1892.0        1892.0  1892.000000   
mean       82.103634                   1.0           1.0     9.460751   
std        59.191488                   0.0           0.0     8.900137   
min        16.474176                   1.0           1.0     2.012581   
25%        59.044904                   1.0           1.0     3.782899   
50%        71.569137                   1.0           1.0     6.400868   
75%        80.914629                   1.0           1.0    11.694187   
max       300.441182                   1.0           1.0    97.297153   

       icustay_seq  label_death_icu  
count       1892.0           1892.0  
mean           1.0              1.0  
std            0.0              0.0  
min            1.0              1.0  
25%            1.0              1.0  
50%            1.0              1.0  
75%            1.0              1.0  
max            1.0              1.0

In [ ]:
demographics_filtered[demographics_filtered["label_death_icu"]==0].describe()

subject_id        hadm_id     icustay_id  los_hospital  \
count  17522.000000   17522.000000   17522.000000  17522.000000   
mean   36444.722577  150125.603698  250415.974717     13.741797   
std    29028.357322   28915.666274   28776.109863     12.476978   
min        3.000000  100001.000000  200003.000000      1.620833   
25%    12971.000000  125203.750000  225500.500000      6.563368   
50%    26246.500000  149936.500000  250841.000000      9.968750   
75%    59825.750000  175505.000000  275289.500000     16.588715   
max    99995.000000  199999.000000  299993.000000    294.660417   

       admission_age  hospital_expire_flag  hospstay_seq       los_icu  \
count   17522.000000          17522.000000       17522.0  17522.000000   
mean       74.205920              0.043945           1.0      6.532633   
std        52.470699              0.204978           0.0      7.276234   
min        15.186897              0.000000           1.0      2.000856   
25%        53.312413              0.000000           1.0      2.833979   
50%        66.518818              0.000000           1.0      3.946233   
75%        78.133566              0.000000           1.0      6.979714   
max       310.280861              1.000000           1.0    153.927975   

       icustay_seq  label_death_icu  
count      17522.0          17522.0  
mean           1.0              0.0  
std            0.0              0.0  
min            1.0              0.0  
25%            1.0              0.0  
50%            1.0              0.0  
75%            1.0              0.0  
max            1.0              0.0

Vital Signs

In [ ]:
final_vital.describe()

icustay_id     heartrate         sysbp        diasbp        meanbp  \
count  3.151743e+06  3.151743e+06  3.151743e+06  3.151743e+06  3.151743e+06   
mean   2.503715e+05  8.659831e+01  1.231295e+02  6.074196e+01  8.015795e+01   
std    2.870068e+04  1.757499e+01  2.300039e+01  1.426881e+01  1.619728e+01   
min    2.000030e+05 -1.000000e+00 -1.000000e+00 -1.000000e+00 -1.000000e+00   
25%    2.255660e+05  7.400000e+01  1.060000e+02  5.100000e+01  6.900000e+01   
50%    2.507610e+05  8.600000e+01  1.210000e+02  5.900000e+01  7.800000e+01   
75%    2.751460e+05  9.800000e+01  1.380000e+02  6.900000e+01  8.950000e+01   
max    2.999930e+05  2.800000e+02  3.550000e+02  2.970000e+02  2.990000e+02   

           resprate         tempc          spo2  label_death_icu  
count  3.151743e+06  3.151743e+06  3.151743e+06     3.151743e+06  
mean   2.038092e+01  3.698067e+01  9.700956e+01     1.333741e-01  
std    6.013303e+00  1.571340e+00  4.579448e+00     3.399787e-01  
min   -1.000000e+00 -1.000000e+00 -1.000000e+00     0.000000e+00  
25%    1.600000e+01  3.650000e+01  9.600000e+01     0.000000e+00  
50%    2.000000e+01  3.700000e+01  9.800000e+01     0.000000e+00  
75%    2.400000e+01  3.755555e+01  9.900000e+01     0.000000e+00  
max    6.900000e+01  4.220000e+01  1.000000e+02     1.000000e+00

Lab Measurements

In [ ]:
final_lab.describe()

icustay_id       albumin            bun      bilirubin  \
count  402774.000000  402774.00000  402774.000000  402774.000000   
mean   250369.434859       1.59474      30.623837       1.184380   
std     28702.320147       1.81297      24.466587       4.374002   
min    200003.000000      -1.00000      -1.000000      -1.000000   
25%    225566.000000      -1.00000      14.000000      -1.000000   
50%    250761.000000       2.30000      23.000000       0.400000   
75%    275146.000000       2.95000      39.000000       1.000000   
max    299993.000000       6.90000     241.000000      82.200000   

             lactate    bicarbonate          bands       chloride  \
count  402774.000000  402774.000000  402774.000000  402774.000000   
mean        1.125841      25.182339       1.297610     104.928369   
std         1.583926       4.924765       5.414536       7.368390   
min        -1.000000      -1.000000      -1.000000      -1.000000   
25%         0.700000      22.000000      -1.000000     101.000000   
50%         1.200000      25.000000      -1.000000     105.000000   
75%         1.700000      28.000000       1.000000     109.000000   
max        29.700000      53.000000      76.000000     155.000000   

          creatinine        glucose     hemoglobin     hematocrit  \
count  402774.000000  402774.000000  402774.000000  402774.000000   
mean        1.364081     131.793220      10.187613      30.290665   
std         1.369498      47.788237       1.694550       4.770681   
min        -1.000000      -1.000000      -1.000000      -1.000000   
25%         0.600000     104.000000       9.100000      27.200000   
50%         0.900000     123.000000      10.000000      29.850000   
75%         1.500000     148.000000      11.100000      32.900000   
max        70.100000    1851.500000      21.100000      67.000000   

            platelet      potassium            ptt         sodium  \
count  402774.000000  402774.000000  402774.000000  402774.000000   
mean      238.101271       4.050497      35.263029     139.151446   
std       148.633425       0.576625      19.201168       7.534353   
min        -1.000000      -1.000000      -1.000000      -1.000000   
25%       137.000000       3.700000      26.400000     136.000000   
50%       207.000000       4.000000      30.300000     139.000000   
75%       302.000000       4.300000      37.500000     142.000000   
max      2292.000000      15.600000     150.000000     184.000000   

                 wbc  label_death_icu  
count  402774.000000    402774.000000  
mean       12.408612         0.132586  
std         7.625251         0.339127  
min        -1.000000         0.000000  
25%         8.300000         0.000000  
50%        11.200000         0.000000  
75%        15.000000         0.000000  
max       528.000000         1.000000

In [ ]:
demographics_filtered.describe()

subject_id        hadm_id     icustay_id  los_hospital  \
count  19414.000000   19414.000000   19414.000000  19414.000000   
mean   36305.652879  149998.389461  250439.306583     13.456947   
std    28966.570968   28917.484208   28773.388017     12.313006   
min        3.000000  100001.000000  200003.000000      0.263889   
25%    12979.000000  124997.500000  225585.250000      6.281250   
50%    26149.500000  149812.000000  250885.500000      9.795139   
75%    59445.750000  175340.750000  275336.750000     16.301910   
max    99995.000000  199999.000000  299993.000000    294.660417   

       admission_age  hospital_expire_flag  hospstay_seq       los_icu  \
count   19414.000000          19414.000000       19414.0  19414.000000   
mean       74.975595              0.137118           1.0      6.817994   
std        53.212971              0.343980           0.0      7.500257   
min        15.186897              0.000000           1.0      2.000856   
25%        53.804782              0.000000           1.0      2.871270   
50%        67.014075              0.000000           1.0      4.070260   
75%        78.496017              0.000000           1.0      7.399604   
max       310.280861              1.000000           1.0    153.927975   

       icustay_seq  label_death_icu  
count      19414.0     19414.000000  
mean           1.0         0.097455  
std            0.0         0.296585  
min            1.0         0.000000  
25%            1.0         0.000000  
50%            1.0         0.000000  
75%            1.0         0.000000  
max            1.0         1.000000

### Write Final Datasets into Postgres

In [ ]:
engine = create_engine('postgresql://mimicuser:mimic3@localhost:5432/mimic3')

#demographics_final.to_sql('demographics_final', engine, if_exists='replace')
final_vital.to_sql('vital_resampled', engine, if_exists='replace')
final_lab.to_sql('lab_resampled', engine, if_exists='replace')

774

### Write Final Datasets into CSV files (alternative to postgres)

In [ ]:
#demographics_final.to_csv(r'.\demographics.csv', index=False)
final_vital.to_csv(r'vitals.csv', index=False)
final_lab.to_csv(r'labs.csv', index=False)